### ISSUE TITLE
Indexing Grammar in combination with vector construction using `[...]` causes ambiguous syntax. 

### ISSUE THREAD

**Issue by sebffischer** [dgkf/R#109]
Because `[[` can be used with a single number like `1` and `[` with a vector like `[1]`, it is not clear what the expression `x[[1]]` represents, i.e. does it call `[[` or `[`.

**Comment by dgkf** [comment]
Yes I know!!! I'm at an impasse with this syntax. This is [described in this issue](https://github.com/dgkf/R/issues/6), but it warrants having it's own dedicated issue.

I _really_ want this syntax to make more sense because I found the `[` vs `[[` syntax really convoluted in R when I was learning and I still see people struggling with it frequently. 

I'd really like to retire `[[` as a separate postfix operator, and instead use dispatch. To me this just feels much more "R" (even if it's not how R works). Making that work raises a few questions:

- How do you differentiate a scalar index from a vector index such that:
  `x[<scalar>]` produces an element of x
  `x[<vector>]` produces a slice of x
- Does this mean we need a scalar type? 
- Even if we did have a scalar type, and `y = 1; x[y]` (`x[<scalar>]`) could be differentiated from `y = [1]; x[y]` (`x[<vector>]`), then this is still the _opposite_ behavior of R. Personally I'm okay with that the syntax can be cleaned up. 
- I think R's roots as an array-programming-language are pretty magical and I wouldn't want to abandon that philosophy just to add scalars.

I don't have good answers to these, so for now it remains a wide-open experiment.

**Comment by dgkf** [comment]
> Agreed! However, maybe we can introduce scalars without moving far away from this idea.

Yeah, this has sort of been what I've been leaning toward, so it's good to hear you reinforce that direction. 

As I think about it more, I think we'll have `Scalar`, `Vector`, and in the future (maybe `Matrix`? and) `NDArray` for higher order arrays. For things that are used frequently like scalars and vectors, I think it's good to specialize the representation and only defer to a more generic representation for higher dimensional.

> This should have some side-benefits regarding performance as we don't have to box all length 1 vectors. 

Certainly  - especially if those basic types all implement `Copy`, then we save a lot on performance and avoiding project overhead by dealing with boxes/shared references. 

---

Cool, well that's about all the confirmation I needed. I think `Scalar` is a necessary feature here.

**Comment by sebffischer** [comment]
Just that I don't forget it: The `AtomicMode` trait should then probably also be renamed, as `Vector`s as are not the atomic data structure anymore once we have scalars.

**Comment by dgkf** [comment]
We might be able to keep it, since it applies to the `T` in `Vector<T>`, not the `Vector` itself.

I imagine that any `T` in either a `Vector<T>` or a `Scalar<T>` will probably need to share some foundational traits, which is currently what `AtomicMode` is for. That said, we might consider renaming it anyways to be more descriptive.

I'm not sure I agree that `Vector` should become `Vector<Scalar<T>>`. In principle I think this would be preferred, but because it's such a widely used data type, we should probably minimize the layers of computation when working with its data.

**Comment by sebffischer** [comment]
> As I think about it more, I think we'll have Scalar, Vector, and in the future (maybe Matrix? and) NDArray for higher order arrays.

I think it would be really cool if we could conceptualize a Scalar as a 0-dimensional array. Not necessarily in the internal representation for which I agree that overhead should be as minimal as possible, but in terms of its semantics. 

> I'm not sure I agree that Vector should become Vector<Scalar<T>>. In principle I think this would be preferred, but because it's such a widely used data type, we should probably minimize the layers of computation when working with its data.

Yes I think that makes sense!

**Comment by dgkf** [comment]
> I think it would be really cool if we could conceptualize a Scalar as a 0-dimensional array.

Yeah, that's how I see it too. I think we'll have distinct data types, but I think the broadcasting should operate such that a 0-dim array is operationally identical, at least for broadcasting/recycling operations like mathematical operators.

**Comment by dgkf** [comment]
Was reading [_Designing Types for R, Empirically_ (Vitek, _et. al._, 2020)](https://dl.acm.org/doi/pdf/10.1145/3428249) today and just wanted to highlight some of their conclusions (_Section 6.1.1_). It's worth a read through all their commentary, but there are at least two that might inform a direction for this project:

## Supporting the inclusion of an explicit _missing_/_na_ 

Similar to `julia`'s `Vector{Union{_, Missing}}`

> [on] _NAs_. Our data supports making the presence of `NA`s explicit. Only 2923 (or 2.78%) of arguments are marked as possibly having `NA`s, thus the overwhelming majority of types appear to be `NA`-free. In practice, programmers check for them and sanitize them if they are present. 

## Supporting the introduction of a Scalar

> [on] _Scalars._ The data also suggests that programmers often use scalars, and do dimensionality checks on their data. In our data 25,064 (or 33.33%) of the arguments are scalar types. While not completely surprising, this is a rather large number. Consider the `hankel.matrix` function, it takes two arguments and checks that `n` is `int`, that `x` is a `vector`, and also, indirectly, that `n` is a a scalar (this comes from the fact that it is used in the guard of a conditional which fails if n is not a scalar).

**Comment by sebffischer** [comment]
thanks so much for sharing this, the title sound fascinating. maybe we could even try to reach out to Jan vitek?
